### DISCUSSION TITLE
What is the ideal way to load balance multiple rtsp-simple-server hosts?

### DISCUSSION THREAD

**Discussion by clee604** [bluenviron/mediamtx#1092] [General]
Hi, many thanks for this project, I've had this running for awhile and it's been working great for my use case. I'm looking for (more) ideal way of load balancing multiple rtsp-simple-server hosts, any thoughts or help would be greatly appreciated.

Use Case:
I am currently using rtsp-simple-server on three linux hosts running docker to proxy my RTSP cameras to clients (mainly TinyCam or VLC); this way there is a single connection to the camera(s) even when there are multiple clients (which can be up to 5-6 at a time). 

Currently manually 'load balance' across the hosts by having the same rtsp-simple-server configuration file and by pointing 'sets' of camera feeds to a host. 

For example there are three linux hosts running rtsp-simple-server in docker:
Host #1 - 10.0.0.1 
Host #2 - 10.0.0.2
Host #3 - 10.0.0.3

6 RTSP cameras:
10.0.0.11
10.0.0.12
10.0.0.13
10.0.0.14
10.0.0.15
10.0.0.16

I would set up TinyCam/VLC clients to access cameras .11 and .12 through host #1, cameras .13 and .14 through host #2 and cameras .15 and .16 through host #3.

This has worked reasonably well except when one of the hosts goes down, then that set of cameras it controls doesn't work unless I manually update the RTSP client IP to use a different host. Also if a lot of clients are accessing both cameras in a set, one host would be under high load while the other hosts are idle.

The host devices are pretty weak, and I may need to add some cameras that need transcoding (ie. ffmpeg) which would greatly increase the load on the devices. Is there a better way for me to load balance between the hosts aside from the manual method I'm doing now?

I saw there feature enhancement under the issues for future Kubernutes scaling support but I'm not knowledgable enough on K9 clusters to know if that would help in my situation (https://github.com/aler9/rtsp-simple-server/issues/998). I would think the ideal state for me is to have a master host that I point my RTSP clients to, and this host coordinates which host actually does the proxy/transcoding as necessary. Is this something I can do using something like docker swarm or I will need to use Kubernutes?

**Comment by aler9** [comment]
Hello, first of all scalability is an open issue that will be dealt with very soon (#998), the problem is that it's more difficult to implement than the one of an HTTP-based service.

In order to scale an HTTP-based service, it's enough to route the request to any of the available nodes, that are usually connected to a database:

```
request --> node1
            node2
            node3
```

This architecture can be applied to rtsp-simple-server if and only if all the nodes own all the video streams, and this is something you don't want, since it would mean routing each video stream N times (where N is the number of nodes) and consuming a bandwidth of `N*stream_bandwidth`.

In order to scale a stream-based service like the one of a media server, a request (for instance, a request to play a stream) must be redirected to the exact node that owns the stream, so we need an additional element (`entry_node`) that analyzes the request and redirects it to the right node:

```
request --> entry_node1 --> node1
            entry_node2     node2
```

You can add an arbitrary number of entry nodes that redirects clients to the right node, each with this configuration:

```yml
paths:
  cam1:
    redirect: rtsp://node1/cam1
  cam2:
    redirect: rtsp://node1/cam2
  cam3:
    redirect: rtsp://node2/cam3
  cam4:
    redirect: rtsp://node2/cam4
```

When a node goes down or you want to replace it, it's enough to update the entry node configuration, for instance by using the API.

**Comment by clee604** [comment]
Big thanks for the explanation!

My setup also relies on the streams being called ondemand, but if I understand the above it shouldn't make much of a difference since each node 'owns' a specific camera. I guess however in this scenario, if a node fails, there wouldn't be a fallback to use a different node for that camera?

**Reply by aler9** [reply]
Fallback nowadays is provided by the orchestrator (Kubernetes or OpenShift). When a node fails, an orchestrator automatically allocates the replica on another node and updates the DNS settings, such that connection to the replica can be restored ASAP. I don't think it's useless to mimic this behavior inside a service since it's provided externally.

**Reply by aler9** [reply]
One additional note: you can either choose a fixed number of replicas, or you can make the number of replicas dynamic and handled by the orchestrator (the auto-scaling feature of Kubernetes). Again, everything is handled by the orchestrator.

What is left to the server is splitting the load between available replicas and redirecting uses to the right replica.

[a comment by an automated account is not shown]
